/*
 * Beautiful Capi generates beautiful C API wrappers for your C++ classes
 * Copyright (C) 2015 Petr Petrovich Petrov
 *
 * This file is part of Beautiful Capi.
 *
 * Beautiful Capi is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Beautiful Capi is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Beautiful Capi.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

/*
 * WARNING: This file was automatically generated by Beautiful Capi!
 * Do not edit this file! Please edit the source API description.
 */

#ifndef EXCEPTION_BADARGUMENT_INCLUDED
#define EXCEPTION_BADARGUMENT_INCLUDED

#include "ExceptionCapi.h"
#include "ExceptionFwd.h"
#include "Exception/Generic.h"

#ifdef __cplusplus

namespace Exception { 

class BadArgument : public Exception::Generic
{
protected:
    void* mObject;
    void SetObject(void* object_pointer)
    {
        mObject = object_pointer;
        if (mObject)
        {
            Exception::Generic::SetObject(exception_badargument_cast_to_base(mObject));
        }
        else
        {
            Exception::Generic::SetObject(0);
        }
    }
public:
    BadArgument(const BadArgument& other) : Exception::Generic(0, false)
    {
        if (other.mObject)
        {
            beautiful_capi_exception_exception_info_t exception_info;
            SetObject(exception_badargument_copy(&exception_info, other.mObject));
            beautiful_capi_Exception::check_and_throw_exception(exception_info.code, exception_info.object_pointer);
        }
        else
        {
            SetObject(0);
        }
    }
    BadArgument(void *object_pointer, bool copy_object) : Exception::Generic(0, false)
    {
        if (object_pointer && copy_object)
        {
            beautiful_capi_exception_exception_info_t exception_info;
            SetObject(exception_badargument_copy(&exception_info, object_pointer));
            beautiful_capi_Exception::check_and_throw_exception(exception_info.code, exception_info.object_pointer);
        }
        else
        {
            SetObject(object_pointer);
        }
    }
    bool IsNull() const
    {
        return !mObject;
    }
    bool IsNotNull() const
    {
        return mObject != 0;
    }
    bool operator!() const
    {
        return !mObject;
    }
    BadArgument* operator->()
    {
        return this;
    }
    const BadArgument* operator->() const
    {
        return this;
    }
    void* Detach()
    {
        void* result = mObject;
        SetObject(0);
        return result;
    }
    BadArgument operator=(const BadArgument& other)
    {
        if (mObject != other.mObject)
        {
            if (mObject)
            {
                exception_badargument_delete(mObject);
                SetObject(0);
            }
            if (other.mObject)
            {
                beautiful_capi_exception_exception_info_t exception_info;
                SetObject(exception_badargument_copy(&exception_info, other.mObject));
                beautiful_capi_Exception::check_and_throw_exception(exception_info.code, exception_info.object_pointer);
            }
            else
            {
                SetObject(0);
            }
        }
        return *this;
    }
    BadArgument() : Exception::Generic(0, false)
    {
        beautiful_capi_exception_exception_info_t exception_info;
        SetObject(exception_badargument_new(&exception_info));
        beautiful_capi_Exception::check_and_throw_exception(exception_info.code, exception_info.object_pointer);
    }
    ~BadArgument()
    {
        if (mObject)
        {
            exception_badargument_delete(mObject);
            SetObject(0);
        }
    }
    const char* GetArgumentName() const
    {
        return exception_badargument_getargumentname(mObject);
    }
};

}

#endif /* __cplusplus */

#endif /* EXCEPTION_BADARGUMENT_INCLUDED */

